scripts/tests/test_zkai_native_attention_mlp_adapter_opening_geometry_budget_gate.py

```
@@ -0,0 +1,173 @@
+    def test_read_json_rejects_symlink(self) -> None:
+        with tempfile.TemporaryDirectory() as tmpdir:
+            directory = pathlib.Path(tmpdir)
+            target = directory / "target.json"
+            symlink = directory / "source.json"
+            target.write_text("{}", encoding="utf-8")
+            try:
+                os.symlink(target, symlink)
+            except (OSError, NotImplementedError):
+                self.skipTest("symlink creation unavailable")
+
+            with self.assertRaisesRegex(gate.AdapterOpeningGeometryBudgetError, "failed to read"):
```

Ruff (0.15.12)

[warning] Use `pytest.raises` instead of unittest-style `assertRaisesRegex`

Replace `assertRaisesRegex` with `pytest.raises`

(PT027)